controllers/flinkcluster_reconciler.go

```diff
@@ -524,57 +520,95 @@ func (reconciler *ClusterReconciler) cancelFlinkJob(jobID string, takeSavepoint
 	var jobStatus = reconciler.observed.cluster.Status.Components.Job
 	var controlStatus = reconciler.observed.cluster.Status.Control
+	var savepointStatus = reconciler.observed.cluster.Status.Savepoint
 
 	if !reconciler.canTakeSavepoint() {
-		return false
+		return false, ""
 	}
 
 	// User requested.
-	if jobSpec.SavepointGeneration > jobStatus.SavepointGeneration {
-		log.Info(
-			"Savepoint is requested",
-			"statusGen", jobStatus.SavepointGeneration,
-			"specGen", jobSpec.SavepointGeneration)
-		return true
-		// Triggered by control annotation
-	} else if controlStatus != nil &&
+
+	// In the case of which savepoint status is in finished state,
+	// savepoint trigger by spec.job.savepointGeneration is not possible
+	// because the field cannot be increased more when savepoint is failed.
+	//
+	// Savepoint retry by annotation is possible because the annotations would be cleared
+	// when the last savepoint was finished and user can attach the annotation again.
+	if controlStatus != nil &&
 		controlStatus.Name == v1beta1.ControlNameSavepoint &&
 		controlStatus.State == v1beta1.ControlStateProgressing {
 		log.Info(
 			"Savepoint is requested",
 			"control name", controlStatus.Name,
 			"control state", controlStatus.State)
-		return true
+		return true, SavepointTriggerReasonUserRequested
+	} else if jobSpec.SavepointGeneration > jobStatus.SavepointGeneration && // TODO: spec.job.savepointGeneration will be deprecated
+		(savepointStatus.State != SavepointStateFailed && savepointStatus.State != SavepointStateTriggerFailed) {
+		log.Info(
+			"Savepoint is requested",
+			"statusGen", jobStatus.SavepointGeneration,
+			"specGen", jobSpec.SavepointGeneration)
+		return true, SavepointTriggerReasonUserRequested
 	}
 
 	if jobSpec.AutoSavepointSeconds == nil {
-		return false
+		return false, ""
 	}
 
 	// First savepoint.
 	if len(jobStatus.LastSavepointTime) == 0 {
-		return true
+		return true, SavepointTriggerReasonScheduled
 	}
 
 	// Interval expired.
 	var tc = &TimeConverter{}
 	var lastTime = tc.FromString(jobStatus.LastSavepointTime)
 	var nextTime = lastTime.Add(
 		time.Duration(int64(*jobSpec.AutoSavepointSeconds) * int64(time.Second)))
-	return time.Now().After(nextTime)
+	return time.Now().After(nextTime), SavepointTriggerReasonScheduled
 }
 
 // Checks whether it is possible to take savepoint.
 func (reconciler *ClusterReconciler) canTakeSavepoint() bool {
 	var jobSpec = reconciler.observed.cluster.Spec.Job
+	var savepointStatus = reconciler.observed.cluster.Status.Savepoint
+	var jobStatus = reconciler.observed.cluster.Status.Components.Job
 	return jobSpec != nil && jobSpec.SavepointsDir != nil &&
-		!reconciler.isJobStopped()
+		!isJobStopped(jobStatus) &&
+		(savepointStatus == nil ||
+			(savepointStatus.State != SavepointStateProgressing) ||
+			(savepointStatus.State == SavepointStateProgressing && savepointStatus.TriggerID == ""))
+}
+
+// Takes savepoint for a job then update job status with the info.
```

Comment:
I don't find where it "then update job status with the info".

Reply:
Changed the comment.